ci/upload-docs.sh

```diff
@@ -0,0 +1,68 @@
+#!/usr/bin/env bash
+#
+# Copyright 2017 Google Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -eu
+
+if [ "${TRAVIS_PULL_REQUEST}" != "false" ]; then
+  echo "Skipping document generation as it is disabled for pull requests."
+  exit 0
+fi
+
+if [ "${GENERATE_DOCS:-}" != "yes" ]; then
+  echo "Skipping document generation as it is disabled for this build."
+  exit 0
+fi
+
+if [ "${TRAVIS_BRANCH:-}" != "master" ]; then
+  echo "Skipping document generation as it is disabled for non-master branches."
+  exit 0
+fi
+
+# The usual way to host documentation in ${GIT_NAME}.github.io/${PROJECT_NAME}
+# is to create a branch (gh-pages) and post the documentation in that branch.
+# We first do some general git configuration:
+
+# Clone the gh-pages branch into the doc/html subdirectory.
+readonly REPO_URL=$(git config remote.origin.url)
+git clone -b gh-pages "${REPO_URL}" doc/html
+
+# Remove any previous content of the branch.  We will recover any unmodified
+# files in a second.
+(cd doc/html && git rm -qfr . || exit 0)
```

Comment:
This is silently masking errors; how about:

```
(cd doc/html && git rm -qfr . || {
  echo "'git rm' resulted in an error; exiting."
  exit 1    # or exit 0? if you insist
})
```

Thoughts?

Reply:
`git-rm` exits with an error code when there are no files match.  Which will happen the first time we create the `gh-pages` branch (anyone of us that forks just "master").  But that got me thinking and the right fix is in.